src/Fetch.js

```diff
@@ -295,303 +57,217 @@ function fetchXHR(fetch, onsuccess, onerror, onprogress, onreadystatechange) {
-    if (!Fetch.xhrs.has(id)) {
-      return;
-    }
-    var ptrLen = (fetchAttrLoadToMemory && fetchAttrStreamData && xhr.response) ? xhr.response.byteLength : 0;
-    var ptr = 0;
-    if (ptrLen > 0 && fetchAttrLoadToMemory && fetchAttrStreamData) {
-#if FETCH_DEBUG
-      dbg(`fetch: allocating ${ptrLen} bytes in Emscripten heap for xhr data`);
-#endif
-#if ASSERTIONS
-      assert(onprogress, 'When doing a streaming fetch, you should have an onprogress handler registered to receive the chunks!');
-#endif
-      // Allocate byte data in Emscripten heap for the streamed memory block (freed immediately after onprogress call)
-      ptr = _malloc(ptrLen);
-      HEAPU8.set(new Uint8Array(/** @type{Array<number>} */(xhr.response)), ptr);
-    }
-    HEAPU32[fetch + {{{ C_STRUCTS.emscripten_fetch_t.data }}} >> 2] = ptr;
-    writeI53ToI64(fetch + {{{ C_STRUCTS.emscripten_fetch_t.numBytes }}}, ptrLen);
-    writeI53ToI64(fetch + {{{ C_STRUCTS.emscripten_fetch_t.dataOffset }}}, e.loaded - ptrLen);
-    writeI53ToI64(fetch + {{{ C_STRUCTS.emscripten_fetch_t.totalBytes }}}, e.total);
-    HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.readyState }}} >> 1] = xhr.readyState;
-    // If loading files from a source that does not give HTTP status code, assume success if we get data bytes
-    if (xhr.readyState >= 3 && xhr.status === 0 && e.loaded > 0) xhr.status = 200;
-    HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.status }}} >> 1] = xhr.status;
-    if (xhr.statusText) stringToUTF8(xhr.statusText, fetch + {{{ C_STRUCTS.emscripten_fetch_t.statusText }}}, 64);
-    onprogress?.(fetch, xhr, e);
-    if (ptr) {
-      _free(ptr);
-    }
-  };
-  xhr.onreadystatechange = (e) => {
-    // check if xhr was aborted by user and don't try to call back
-    if (!Fetch.xhrs.has(id)) {
-      {{{ runtimeKeepalivePop() }}}
-      return;
-    }
-    HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.readyState }}} >> 1] = xhr.readyState;
-    if (xhr.readyState >= 2) {
-      HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.status }}} >> 1] = xhr.status;
-    }
-    onreadystatechange?.(fetch, xhr, e);
-  };
-#if FETCH_DEBUG
-  dbg(`fetch: xhr.send(data=${data})`);
-#endif
-  try {
-    xhr.send(data);
-  } catch(e) {
-#if FETCH_DEBUG
-    dbg(`fetch: xhr failed with exception: ${e}`);
-#endif
-    onerror?.(fetch, xhr, e);
+  if(userName && password) {
+    var userNameStr = userName ? UTF8ToString(userName) : undefined;
+    var passwordStr = password ? UTF8ToString(password) : undefined;
+
+    headers.set('Authorization', 'Basic ' + btoa(userNameStr + ":" + passwordStr));
   }
-}
 
-function startFetch(fetch, successcb, errorcb, progresscb, readystatechangecb) {
-  // Avoid shutting down the runtime since we want to wait for the async
-  // response.
-  {{{ runtimeKeepalivePush() }}}
+  var body = (dataPtr && dataLength) ? HEAPU8.slice(dataPtr, dataPtr + dataLength) : null;
+
+  const controller = new AbortController();
```

Comment:
We should probably check that this is supported on all browser we support targetting.